### PR TITLE
Add a complete implementation of the FreeRTOS semaphore APIs.

### PR DESCRIPTION
**Pull request by davidchisnall**
FreeRTOS combines various kinds of mutex and semaphore under the semaphore umbrella.  This is somewhat confusing because recursive mutexes are supported by a completely different set of APIs.

This implementation lets individual components opt into whichever once they need.  If they need only one, then we avoid space and instruction overheads of dynamic dispatch.

### CODE REVIEW

File: sdk/include/FreeRTOS-Compat/semphr.h [hunk 1]
```diff
@@ -53,34 +178,272 @@ xSemaphoreCreateCounting(UBaseType_t uxMaxCount, UBaseType_t uxInitialCount)
...
+	}
 	return ret == 0;
 }
 
+/**
+ * Take a semaphore from code that may run in an ISR.  CHERIoT RTOS does not
+ * permit code to run in ISRs and so this simply calls the normal code paths.
+ *
+ * The second parameter is a pointer to a variable that will be set to true if
+ * the caller needs to yield to wake a higher-priority task.
```
Comment (nwf-msr) [inline sdk/include/FreeRTOS-Compat/semphr.h:338]: Not yet supported?

File: sdk/include/FreeRTOS-Compat/semphr.h [hunk 2]
```diff
@@ -53,34 +178,272 @@ xSemaphoreCreateCounting(UBaseType_t uxMaxCount, UBaseType_t uxInitialCount)
...
  */
 __always_inline static inline void
 vSemaphoreDelete(SemaphoreHandle_t xSemaphore)
 {
 	free(xSemaphore);
 }
 
+/**
+ * Get the current owner (thread ID) of a mutex.  If this is a counting
+ * semaphore, or if the lock is not held, returns zero.
```
Comment (nwf-msr) [inline sdk/include/FreeRTOS-Compat/semphr.h:276]: This probably deserves the same comment about racing as below, with the (anti)caveat that an answer of "this thread" _is_ stable.

File: sdk/include/FreeRTOS-Compat/semphr.h [hunk 3]
```diff
@@ -44,7 +169,7 @@ __always_inline static inline SemaphoreHandle_t
 xSemaphoreCreateCounting(UBaseType_t uxMaxCount, UBaseType_t uxInitialCount)
 {
 	SemaphoreHandle_t semaphore =
-	  (SemaphoreHandle_t)malloc(sizeof(struct CountingSemaphoreState));
+	  (SemaphoreHandle_t)malloc(sizeof(SemaphoreHandle_t));
```
Comment (nwf-msr) [inline sdk/include/FreeRTOS-Compat/semphr.h:172]: Do you really mean `sizeof(SemaphoreHandle_t)`?

```suggestion
	  (SemaphoreHandle_t)malloc(sizeof(StaticSemaphore_t));
```

File: sdk/include/FreeRTOS-Compat/semphr.h [hunk 4]
```diff
@@ -1,18 +1,141 @@
...
+#if !defined(CHERIOT_EXPOSE_FREERTOS_SEMAPHORE) &&                             \
+  !defined(CHERIOT_EXPOSE_FREERTOS_MUTEX) &&                                   \
+  !defined(CHERIOT_EXPOSE_FREERTOS_RECURSIVE_MUTEX)
+#	error At least one out of CHERIOT_EXPOSE_FREERTOS_SEMAPHORE, \
+  CHERIOT_EXPOSE_FREERTOS_MUTEX, or CHERIOT_EXPOSE_FREERTOS_RECURSIVE_MUTEX \
+  must be defined.
+#endif
+
+#ifdef CHERIOT_EXPOSE_FREERTOS_SEMAPHORE
+#	define CHERIOT_FREERTOS_SEMAPHORE 1
```
Comment (nwf-msr) [inline sdk/include/FreeRTOS-Compat/semphr.h:29]: While I don't object to these `EXPOSE`-less `define`s, are they doing anything that `defined(CHERIOT_EXPOSE_...)` can't?

File: sdk/include/FreeRTOS-Compat/semphr.h [hunk 5]
```diff
@@ -53,34 +178,272 @@ xSemaphoreCreateCounting(UBaseType_t uxMaxCount, UBaseType_t uxInitialCount)
...
+	return pxMutexBuffer;
+}
+
+/**
+ * Create a heap-allocated mutex.
+ */
+__always_inline static inline SemaphoreHandle_t xSemaphoreCreateMutex()
+{
+	SemaphoreHandle_t semaphore =
+	  (SemaphoreHandle_t)malloc(sizeof(SemaphoreHandle_t));
```
Comment (nwf-msr) [inline sdk/include/FreeRTOS-Compat/semphr.h:222]: As above?

File: sdk/include/FreeRTOS-Compat/semphr.h [hunk 6]
```diff
@@ -53,34 +178,272 @@ xSemaphoreCreateCounting(UBaseType_t uxMaxCount, UBaseType_t uxInitialCount)
...
+}
+
+/**
+ * Create a heap-allocated recursive mutex.
+ */
+__always_inline static inline SemaphoreHandle_t
+xSemaphoreCreateRecursiveMutex(void)
+{
+	SemaphoreHandle_t semaphore =
+	  (SemaphoreHandle_t)malloc(sizeof(SemaphoreHandle_t));
```
Comment (nwf-msr) [inline sdk/include/FreeRTOS-Compat/semphr.h:253]: Third time's the charm?

**Comment by davidchisnall** [comment]
Should all be fixed and I've added a compile test for several of the variants.

File: sdk/include/FreeRTOS-Compat/semphr.h (hunk 4 above)
Comment (nwf-msr) [reply sdk/include/FreeRTOS-Compat/semphr.h:29]: FWIW, `defined()` returns `1` or `0`, so I think you can `defined() + defined() + ...`

Comment (davidchisnall) [reply sdk/include/FreeRTOS-Compat/semphr.h:29]: Ah, I forgot that C thinks Booleans are integers. That should simplify it slightly.

Comment (davidchisnall) [reply sdk/include/FreeRTOS-Compat/semphr.h]: Fixed, hopefully ready to merge now.

**Review by nwf-msr — APPROVED** [review]

**Pull request merged**